controller: unit tests for overlapping and recreate deployments

0xmichalis opened the pull request:
Belated unit tests for https://github.com/kubernetes/kubernetes/pull/38080 and https://github.com/kubernetes/kubernetes/pull/36748.

@kubernetes/sig-apps-misc

0xmichalis committed:
controller: unit tests for overlapping and recreate deployments

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39088)
<!-- Reviewable:end -->

kargakis commented:
Darn, can't repro locally this one. Will have a closer look tomorrow.

kargakis commented:
@k8s-bot kubemark e2e test this https://github.com/kubernetes/kubernetes/issues/38163

smarterclayton commented:
Tests look good to me but will defer to @janetkuo for LGTM

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	foo.CreationTimestamp = now
+	bar := newDeployment("bar", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	bar.CreationTimestamp = later
```
Should we use `SelectorUpdateAnnotation` instead of `CreationTimestamp`? (We can keep CreationTimestamp in this test, but make the older one's selector updated last)

kargakis replied:
`SelectorUpdateAnnotation` should be tested in the util package but it seems we have no unit tests there? I will add one there.

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	foo.CreationTimestamp = now
+	bar := newDeployment("bar", 1, nil, nil, nil, map[string]string{"foo": "bar"})
```
Suggest making selector not equal but overlap, to catch the bug we had earlier

kargakis replied:
ok

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+	actions := f.client.Actions()
+	d := actions[0].(core.UpdateAction).GetObject().(*extensions.Deployment)
+	if len(d.Annotations[util.OverlapAnnotation]) == 0 {
+		t.Errorf("annotations werent updated for the overlapping deployment: %v", d.Annotations)
```
Suggest also verifying the content of the overlapping annotation

janetkuo replied:
Suggest making this (check overlapping annotation and print errors) a function so that we can reuse it

kargakis replied:
ok for checking the content - not sure where else the function is going to be reduced so I am leaving it out for now.

kargakis replied:
ok for checking the content - not sure where else the function is going to be reused so I am leaving it out for now.

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	foo.CreationTimestamp = now
+	bar := newDeployment("bar", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	bar.CreationTimestamp = later
```
Same suggestions here regarding selectors and CreationTimestamp

janetkuo replied:
If this is changed, comments need to be updated as well

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+// one will be synced and the newer will be marked as overlapping. Note that in reality it's
+// not always the older deployment that is the one that works vs the rest but the one which
+// has the selector unchanged for longer time.
+func TestSyncNonOverlappingDeployment(t *testing.T) {
```
NonOverlapping may be a bit misleading. Perhaps "Overlapped"?

janetkuo replied:
Actually, can't we combine `TestOverlappingDeployments` and `TestSyncNonOverlappingDeployment` as a single test?

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+	f.run(getKey(foo, t))
+
+	for _, a := range f.client.Actions() {
+		action, ok := a.(core.UpdateAction)
```
Remove this and just do `filterInformerActions` on `f.client.Actions()`?

janetkuo replied:
After it's filtered, there should be only update actions. Then you can just check d.Name and d.Annotations.

kargakis replied:
I am not gaining really anything by doing this. I still have a slice of actions that I need to traverse, cast to UpdateAction, get the object, and check the annotation.

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+		}
+
+		if len(d.Annotations[util.OverlapAnnotation]) > 0 {
+			t.Errorf("annotations werent updated for the overlapping deployment: %v", d.Annotations)
```
weren't cleaned up (just to differentiate with the other check)

kargakis replied:
ok

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+}
+
+// generatePodFromRS creates a pod, with the input ReplicaSet's selector and its template
+func generatePodFromRS(rs *extensions.ReplicaSet) *v1.Pod {
```
nit: move it below its callers

kargakis replied:
ok

janetkuo reviewed (commented):
I have a few comments

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+	now := metav1.Now()
+	later := metav1.Time{Time: now.Add(time.Minute)}
+
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
```
Optional: pass something like `DeploymentOptions` to `newDeployment` instead of 6 parameters.

kargakis replied:
This is a significant test refactor that I wasn't planning on undertaking here.

janetkuo replied:
agree

janetkuo replied:
agree

0xmichalis committed:
Fix and tests for SelectorUpdatedBefore

kargakis commented:
Two bugs that I found out while adding a test for the SelectorUpdatedAnnotation:
* time.Parse input was wrong so we essentially never parsed correctly the SelectorUpdatedAnnotation, thus always falling back to the creation timestamp. I have fixed this.
* we still have an issue with some selector updates (look out for the TODO comment in one of the test cases)

kargakis commented on pkg/controller/deployment/util/deployment_util_test.go:
```diff
@@ -1107,3 +1107,117 @@ func TestDeploymentTimedOut(t *testing.T) {
+			// d1 is created first, d2 is created after and its selector overlaps
+			// with d1. d2 is marked as overlapping correctly. If d1's selector is
+			// updated and continues to overlap with the selector of d2 then d1 is
+			// now marked overlapping and d2 is cleaned up. I think this is a bug,
```
We can simply look for the overlapping annotation between two deployments prior to comparing their timestamps and if one of them has it and points to the other, then we can skip the timestamp check.

janetkuo replied:
good catch

janetkuo commented on pkg/controller/deployment/util/deployment_util_test.go:
```diff
@@ -1107,3 +1107,117 @@ func TestDeploymentTimedOut(t *testing.T) {
+			// the concept is proved by the following test.
+			name: "d1 created before d2 but updated its selector",
+
+			d1:               generateDeployment("foo"),
```
`creationTimestamp1: &now`

janetkuo commented on pkg/controller/deployment/util/deployment_util_test.go:
```diff
@@ -1107,3 +1107,117 @@ func TestDeploymentTimedOut(t *testing.T) {
+	}
+}
+
+func addTimestamp(d extensions.Deployment, timestamp metav1.Time) extensions.Deployment {
```
`addTimestamp` and `addSelectorUpdatedAnnotation` are never called?

janetkuo reviewed (commented):
Thanks for the good catch. Only a few nits.

kargakis commented:
@janetkuo I went ahead and fixed the bug about the older deployment flapping to overlapping once its selector is updated but still overlaps and added an additional test case that covers it (see TestSelectorUpdate in deployment_controller_test.go). ptal

0xmichalis committed:
Fix the overlapping e2e test for deployments

Fix the test to correctly expect the older deployment to be marked as
overlapping since that has the newest selector that overlaps.

kargakis commented:
@k8s-bot unit test this

kargakis commented:
@k8s-bot unit test this

kargakis commented:
@janetkuo ptal

janetkuo commented:
Review status: 0 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 612 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3UVfJ8R-U5BFqE1vL:-Ka3UVfKLLtTTbcjiku9:b-wzzs0t) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/deployment_controller.go#L612)):*
> ```Go
> 		// If the otherD deployment overlaps with the current we need to identify which one
> 		// holds the set longer and mark the other as overlapping. Requeue the overlapping
> 		// deployments if this one has been marked deleted, we only update its status as long
> ```

We need to update comments  

---

*[pkg/controller/deployment/deployment_controller_test.go, line 377 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-KZX3_uk0gPUAuf_UwtY-r1-377:-K_kSzvW8PDANd1hF-nz:bcak7ld) ([raw file](https://github.com/kubernetes/kubernetes/blob/5d71c090c99c0b2f87e4f0f3baa3e99e931e6b14/pkg/controller/deployment/deployment_controller_test.go#L377)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary><blockquote>

I am not gaining really anything by doing this. I still have a slice of actions that I need to traverse, cast to UpdateAction, get the object, and check the annotation.
</blockquote></details>

Yeah, I was thinking that if you filtered these actions first, they should only be update deployments and the `ok` check wouldn't be necessary. 

---

*[pkg/controller/deployment/deployment_controller_test.go, line 227 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3WBzRyStif758gYwy:-Ka3WBzRyStif758gYwz:bevcth1) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/deployment_controller_test.go#L227)):*
> ```Go
> 
> 	f.expectCreateRSAction(rs)
> 	f.expectUpdateDeploymentStatusAction(d)
> ```

From this I realized that `f.run` doesn't check subresources. Would you fix that also?

---

*[pkg/controller/deployment/deployment_controller_test.go, line 385 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3X2vKMbAgcNl6_mHl:-Ka3X2vKMbAgcNl6_mHm:b8kf7qj) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/deployment_controller_test.go#L385)):*
> ```Go
> 
> 		if len(d.Annotations[util.OverlapAnnotation]) > 0 {
> 			t.Errorf("deployment %q should not have the overlapping annotation", d.Name)
> ```

Maybe check that `bar` still has overlapping annotation

---

*[pkg/controller/deployment/util/deployment_util.go, line 1007 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XGdLiNR7E2KcwIPL:-Ka3XGdLiNR7E2KcwIPM:b-9p5jqq) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util.go#L1007)):*
> ```Go
> // SelectorUpdatedBefore returns true if the former deployment's selector
> // is updated before the latter, false otherwise. Callers of this function
> // should first check for the existence of the overlapping annotation in
> ```

IMO this additional comment isn't necessary since this function only checks if d1's selector is updated before d2, and doesn't care about overlapping deployments 

---

*[pkg/controller/deployment/util/deployment_util_test.go, line 1158 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XaQCK9kiGL2gDmo_:-Ka3XaQCK9kiGL2gDmoa:bade3g6) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util_test.go#L1158)):*
> ```Go
> 			// updated and continues to overlap with the selector of d2 then d1 is
> 			// now marked overlapping and d2 is cleaned up. Proved by the following
> 			// test case. Callers of SelectorUpdatedBefore should first check for
> ```

Question about the comment here: does this test prove this bug? IIUC the test will pass before the latest overlapping annotation change, which means this test doesn't prove it?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

janetkuo reviewed (commented):
Reviewed in Reviewable instead

kargakis commented:
*[pkg/controller/deployment/deployment_controller.go, line 612 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3UVfJ8R-U5BFqE1vL:-Ka6k4MZMNw4WTbt7-gZ:b-caplmu) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/deployment_controller.go#L612)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

We need to update comments  
</blockquote></details>

Not sure what needs to be updated in this comment, it still holds true.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
*[pkg/controller/deployment/deployment_controller_test.go, line 227 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3WBzRyStif758gYwy:-Ka6kMRxnodq9Q87yg0N:babmf8e) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/deployment_controller_test.go#L227)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

From this I realized that `f.run` doesn't check subresources. Would you fix that also?
</blockquote></details>

This PR has already been dragged to fix more than the original intention. I prefer to open an issue about this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
*[pkg/controller/deployment/deployment_controller_test.go, line 385 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3X2vKMbAgcNl6_mHl:-Ka6kvcdZsvp88cXIcX1:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/deployment_controller_test.go#L385)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

Maybe check that `bar` still has overlapping annotation
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
*[pkg/controller/deployment/util/deployment_util.go, line 1007 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XGdLiNR7E2KcwIPL:-Ka6lHLUcqW_kazsLF1j:b-hfpv0l) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util.go#L1007)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

IMO this additional comment isn't necessary since this function only checks if d1's selector is updated before d2, and doesn't care about overlapping deployments 
</blockquote></details>

removed

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
*[pkg/controller/deployment/util/deployment_util_test.go, line 1158 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XaQCK9kiGL2gDmo_:-Ka6l_H5dy2jNTqSkSQX:b-kksa4k) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util_test.go#L1158)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

Question about the comment here: does this test prove this bug? IIUC the test will pass before the latest overlapping annotation change, which means this test doesn't prove it?
</blockquote></details>

I don't understand what you mean by "before the latest overlapping annotation change". `LastSelectorUpdate` which is what `SelectorUpdatedBefore` calls first checks for the existence of the annotation that holds the selector update and then falls back to the creation timestamp. d1 is created before d2 but since its selector is updated after d2 is created it will always be considered as the overlapping deployment which is wrong.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/issues/39604

mbohlool commented:
@k8s-bot unit test this #39604

janetkuo commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/controller/deployment/util/deployment_util_test.go, line 1158 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XaQCK9kiGL2gDmo_:-Ka8gWfEHZ9pgObK2-Cg:b-v1bald) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util_test.go#L1158)):*
> "before the latest overlapping annotation change"

This means the last commit / patch you add 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/controller/deployment/util/deployment_util_test.go, line 1158 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XaQCK9kiGL2gDmo_:-Ka8hIEoE8QVHLun2jUi:b-b3ijp0) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util_test.go#L1158)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

> "before the latest overlapping annotation change"

This means the last commit / patch you add 
</blockquote></details>

To be more clear, my question was: does this test catch the bug (described in the scenario) you're fixing?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
*[pkg/controller/deployment/util/deployment_util_test.go, line 1158 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-Ka3XaQCK9kiGL2gDmo_:-KaBiRulVdxjbBv-K9sr:b-35t2zz) ([raw file](https://github.com/kubernetes/kubernetes/blob/669b5ca4b5c784da526e31450ba83041f7493574/pkg/controller/deployment/util/deployment_util_test.go#L1158)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary><blockquote>

To be more clear, my question was: does this test catch the bug (described in the scenario) you're fixing?
</blockquote></details>

The bug is fixed by my last commit but the main reason this function exists is to resolve overlaps (actually I am not aware of any other reason we would want it) and the unit test proves that you can't know which Deployment holded the set for longer.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088)*
<!-- Sent from Reviewable.io -->

kargakis commented:
@k8s-bot unit test this

janetkuo commented:
/lgtm

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39088#-:-KaDsEyOMUuSXXmIwgU7:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

mikedanese commented:
Deployment tests are all the sudden very flaky.

https://github.com/kubernetes/kubernetes/pull/37039#issuecomment-272043352
https://github.com/kubernetes/kubernetes/pull/39771#issuecomment-272036911

mikedanese commented:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/37039/pull-kubernetes-unit/12750/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39771/pull-kubernetes-unit/12741/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39771/pull-kubernetes-unit/12738
